mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -190,59 +189,69 @@ class StringIndexer @Since("1.4.0") (
         new Column(If(col.isNaN.expr, Literal(null), col.expr)).cast(StringType)
       }
     }
+  }
 
+  private def countByValue(
+      dataset: Dataset[_],
+      inputCols: Array[String]): Array[OpenHashMap[String, Long]] = {
+
+    val aggregator = new StringIndexerAggregator(inputCols.length)
+    implicit val encoder = Encoders.kryo[Array[OpenHashMap[String, Long]]]
+
+    val selectedCols = getSelectedCols(dataset, inputCols)
     dataset.select(selectedCols: _*)
       .toDF
       .groupBy().agg(aggregator.toColumn)
       .as[Array[OpenHashMap[String, Long]]]
       .collect()(0)
   }
 
-  @Since("2.0.0")
-  override def fit(dataset: Dataset[_]): StringIndexerModel = {
-    transformSchema(dataset.schema, logging = true)
+  private def sortByFreq(dataset: Dataset[_], ascending: Boolean): Array[Array[String]] = {
+    val (inputCols, _) = getInOutCols()
 
+    val sortFunc = StringIndexer.getSortFunc(ascending = ascending)
+    val orgStrings = countByValue(dataset, inputCols).toSeq
+    ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
+      counts.toSeq.sortWith(sortFunc).map(_._1).toArray
+    }.toArray
+  }
+
+  private def sortByAlphabet(dataset: Dataset[_], ascending: Boolean): Array[Array[String]] = {
     val (inputCols, _) = getInOutCols()
 
     // If input dataset is not originally cached, we need to unpersist it
     // once we persist it later.
     val needUnpersist = dataset.storageLevel == StorageLevel.NONE
 
+    dataset.persist()
```

Comment:
This persist, I think, was because there was many same operations per column. As we finish Spark job once for all columns, maybe we don't need persist now?